src/merkle_tree.rs

```diff
@@ -22,100 +24,116 @@ impl MerkleTree {
-		     if left_n_right.get(1).is_none() {
-		        // This is the case where we have an uneven
-		        // amount of nodes
-		         hash(&[left_n_right[0], left_n_right[0]])
-		     } else {
-		         hash(&[left_n_right[0], left_n_right[1]])
-		     }
-		 }
-	      )
-	      .collect();
-	  tree.push(new_leaves.clone());
-	  if new_leaves.len() == 1 {
-	      break;
-	  }
+            new_leaves = new_leaves
+                .chunks(2)
+                .map(|left_n_right| {
+                    if left_n_right.get(1).is_none() {
+                        // This is the case where we have an uneven
+                        // amount of nodes
+                        hash(&[left_n_right[0], left_n_right[0]])
+                    } else {
+                        hash(&[left_n_right[0], left_n_right[1]])
+                    }
+                })
+                .collect();
+            tree.push(new_leaves.clone());
+            if new_leaves.len() == 1 {
+                break;
+            }
         }
 
         MerkleTree { leaves: tree }
-
     }
 
-    pub fn generate_proof<T: std::convert::AsRef<[u8]>>(&self, elem: &T) -> Option<Vec<HashedData>> {
-
+    // This function returns the index of the hash mainly to make
+    // debugging easier. Plus, I thinks it's a cool bonus. It can
+    // always be ignored with (_, hash)
+    pub fn generate_proof<T: std::convert::AsRef<[u8]>>(
+        &self,
+        elem: &T,
+    ) -> Option<Vec<(Position, HashedData)>> {
         // It there is no first level, for whatever reason, reaturn None
         let first_level = self.leaves.get(0)?;
 
         // If the requested item is not present in the data block, we
         // also return None
-        let mut index = first_level
-	  .iter()
-	  .position(|og_data| {
-	      let check: HashedData = Sha3_256::digest(elem).into();
-	      *og_data == check
-	  })?;
-
-        // We need the brother of the requested item.
-        let brother = if index % 2 == 0 { index + 1 } else { index - 1 };
-        let mut needed_values: Vec<_> = vec![index, brother];
-
-        // The first two values *MUST* be sorted, since the hashing
-        // function varies output depending on order.
-        needed_values.sort();
-
-        // NOTE: This section is very imperative, I'd like to simplify
-        // it later, for now it works. But I am not a fan at all.
-        // This is just a WIP
-        for _ in self.leaves.iter().skip(1) {
-	  let parent_index = (index as f32 / 2.0).floor() as usize;
-	  let uncle_index = if parent_index % 2 == 0 { parent_index + 1 } else { parent_index - 1 };
-
-	  index = parent_index;
-	  needed_values.push(uncle_index);
+        let mut index = first_level.iter().position(|og_data| {
+            let check: HashedData = Sha3_256::digest(elem).into();
+            *og_data == check
+        })?;
+
+        let first_hash = self.leaves[0][index];
+        let mut hashes: Vec<(Position, HashedData)> = vec![(index, first_hash)];
```

Comment:
You don't need to include the searched hash in the proof vector. It seems to me that this is going to generate a problem in the `verify()` since you are going to be hashing more.

Reply:
Thanks! That's probably what's wrong. I'll remove it.

Reply:
That indeed fixed it! You're a life saver Dami ^_^